ASSERTION FAILED: checkedPtrCountWithoutThreadCheck() in SQLiteStorageArea::prepareDatabase

szewai opened the pull request:
#### b290846137ed3270ffa41ef764d3452e21b5f206
<pre>
ASSERTION FAILED: checkedPtrCountWithoutThreadCheck() in SQLiteStorageArea::prepareDatabase
<a href="https://bugs.webkit.org/show_bug.cgi?id=300294">https://bugs.webkit.org/show_bug.cgi?id=300294</a>
<a href="https://rdar.apple.com/162083739">rdar://162083739</a>

Reviewed by Chris Dumez.

SQLiteStorageArea::prepareDatabase might set m_database to null after m_database is captured in a local CheckedPtr
variable resultDatabase. When the CheckedPtr variable is destroyed at function exit, the assertion will be hit as the
SQLiteDatabase object is already gone. To fix this, this patch uses a local UniqueRef variable to replace the
CheckedPtr. The UniqueRef is only moved to m_database if operation succeeds, so prepareDatabase does not need to reset
m_database in multiple conditions.

This patch also contains a drive-by fix that when prepareDatabase tries to open database for the second time, it does
not pass the right open options.

Canonical link: <a href="https://commits.webkit.org/301155@main">https://commits.webkit.org/301155@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/bf1d24117aff5a1b05e4f189d42bffd454e48ebb

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/125094 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/44761 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/35500 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/131942 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/76967 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/206745d0-d4c4-4d66-aab2-4b1fc3c8fcd8) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/45452 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/53326 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/95227 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/76967 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/c0ea3977-0318-4934-afa1-6401eaa83b1a) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/128048 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/36283 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/111874 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/75769 "Passed tests") | | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/7e5e0ba3-1b43-4a30-979c-9cfcbba2377c) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/35192 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/75423 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/106048 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/30263 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/134621 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/51908 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/39705 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/103696 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/52338 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/108091 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/103468 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26343 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/48809 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/27106 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/48961 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/51797 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/57587 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/51167 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/54523 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/52858 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/NetworkProcess/storage/SQLiteStorageArea.cpp:
```diff
@@ -164,36 +164,45 @@ bool SQLiteStorageArea::createTableIfNecessary()
     return true;
 }
 
+static Expected<UniqueRef<WebCore::SQLiteDatabase>, int> createAndOpenDatabase(const String& path, bool createParentDirectory = false)
```
We should not be having boolean parameters like these, it makes call sites less readable. Please use an enum class.

cdumez reviewed (approved):
LGTM but please move to a enum class parameter instead of the boolean.

szewai committed:
ASSERTION FAILED: checkedPtrCountWithoutThreadCheck() in SQLiteStorageArea::prepareDatabase
https://bugs.webkit.org/show_bug.cgi?id=300294
rdar://162083739

Reviewed by Chris Dumez.

SQLiteStorageArea::prepareDatabase might set m_database to null after m_database is captured in a local CheckedPtr
variable resultDatabase. When the CheckedPtr variable is destroyed at function exit, the assertion will be hit as the
SQLiteDatabase object is already gone. To fix this, this patch uses a local UniqueRef variable to replace the
CheckedPtr. The UniqueRef is only moved to m_database if operation succeeds, so prepareDatabase does not need to reset
m_database in multiple conditions.

This patch also contains a drive-by fix that when prepareDatabase tries to open database for the second time, it does
not pass the right open options.

Canonical link: https://commits.webkit.org/301155@main
